fabric-rendering-fluids-v1/src/main/java/net/fabricmc/fabric/api/client/render/fluid/v1/CustomFluidRenderer.java

```diff
@@ -0,0 +1,38 @@
+package net.fabricmc.fabric.api.client.render.fluid.v1;
+
+import net.fabricmc.fabric.impl.client.rendering.fluid.DefaultFluidRenderer;
+import net.minecraft.client.render.VertexConsumer;
+import net.minecraft.client.render.block.FluidRenderer;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.BlockRenderView;
+
+/**
+ * Interface to handle custom rendering for fluids, if providing custom fluid sprites is not enough. This interface
+ * allows you to tessellate your custom fluids yourself and will override any {@link FluidRenderHandler} registered for
+ * your fluids.
+ */
+public interface CustomFluidRenderer {
+	/**
+	 * The default fluid renderer, which delegates to vanilla's fluid renderer.
+	 */
+	CustomFluidRenderer DEFAULT = DefaultFluidRenderer.INSTANCE;
+
+	/**
+	 * Render your fluid. This method will be invoked instead of the default fluid renderer ({@link FluidRenderer}), and
+	 * hence overrides any {@link FluidRenderHandler}s registered for the fluid this renderer is registered for.
+	 *
+	 * <p>Even though the {@link FluidRenderHandler} for your fluid is not used, you can decide to manually use it by
+	 * loading it from the {@link FluidRenderRegistry}.
+	 *
+	 * <p>Note that this method must *only* return {@code true} if at least one face is tessellated. If no faces are
+	 * tessellated this method must return {@code false}.
+	 *
+	 * @param pos            The position in the world, of the fluid to render
+	 * @param world          The world the fluid is in
+	 * @param vertexConsumer The vertex consumer to tessellate the fluid in
+	 * @param state          The fluid state being rendered
+	 * @return Whether anything is tessellated
+	 */
+	boolean renderFluid(BlockPos pos, BlockRenderView world, VertexConsumer vertexConsumer, FluidState state);
```

Comment:
Is there any reason this is not a method in `FluidRenderHandler`? Something like

```java
default boolean renderFluid(BlockPos pos, BlockRenderView world, VertexConsumer vertexConsumer, FluidState state) {
    // static method that delegates to vanilla, somewhere in impl.
    return FluidRenderHandlerImpl.defaultRenderFluid(pos, world, vertexConsumer, state);
}
```

would work as well I think?

Reply:
Maybe you should pass the `FluidRenderer` as a parameter? Then the default method body could just call it directly.

Reply:
Hm... I think the delegation via `FluidRenderHandler` is more stable than directly exposing `FluidRenderer`. Can look into that, but I also added a `DefaultLikeFluidRenderer` now, that imitates vanilla's behaviour but can be inherited to modify certain parameters or behaviours.

Reply:
Although, I am not sure how well it works to ship a `DefaultLikeFluidRenderer` that is an entire copy of the default fluid renderer, except with some extra methods to inherit.

Reply:
This is probably not a good idea due to the maintenance cost that comes with it.

Reply:
I added access to the default implementation via an extra parameter that wraps the `FluidRenderer`, to not directly expose the `FluidRenderer` (because I'm not sure how robust that is, plus it's not documented)

Reply:
The new `CustomFluidRenderer` interface was intended by the way. This makes it possible to define one render type for fluids with different render handlers.

I can understand the maintenance cost of `DefaultLikeFluidRenderer`. I am fine with removing it if mainenance cost outweighs its usefullness.
However, it could be handy if a lot of modders just want to adjust their fluids more closely, while keeping it look like vanilla, without having to imitate vanilla's code all by themselves or break Fabric's mixins with their own. Plus you avoid a whole bunch of mixins from Fabric that could drain some performance. I might even believe it's not even that hard to maintain, but I'm not sure.
Also, I realised it's in the `impl` package :facepalm: